src/server/routers/session.ts

```diff
@@ -22,42 +31,84 @@ const defaultSessionSelect = Prisma.validator<Prisma.SessionSelect>()({
 
 });
 
+const settingsObject = z.object({
+  use_images: z.boolean(),
+  use_nsfw: z.boolean(),
+
+  timer_multiplier: z.number().gte(0).lte(2),
+  turn_multiplier: z.number().gte(0).lte(2),
+  backlog_percentage: z.number().gte(0).lte(0.95),
+});
+
 export const sessionRouter = router({
   // GET
   get: procedure
     .input(
       z.object({
-        id: z.string().length(24),
+        id: DbId,
       }),
     )
     .query(async ({ input }) => {
       const { id } = input;
-      const addon = await prisma.session.findUnique({
-        where: { id },
+
+      const session = await prisma.session.findUnique({
         select: defaultSessionSelect,
+        where: { id },
       });
-      if (!addon) {
+
+      if (!session) {
         throw new TRPCError({
           code: "NOT_FOUND",
           message: `No session with id '${id}'`,
         });
       }
-      return addon;
+
+      return session;
     }),
+
   // ADD
   add: procedure
     .input(
       z.object({
-        addon_ids: z.string().length(24).array(),
-        game_id: z.string().length(24),
+        settings: settingsObject,
+        game_id: DbId,
       }),
     )
     .mutation(async ({ input }) => {
-      const addon = await prisma.session.create({
-        data: { ...input, settings: {}, join_code: "" },
-        select: defaultSessionSelect,
+      const { game_id, settings } = input;
+
+      // Generate x random codes to try.
+      const randomCode = () => ((Math.random() + 1).toString(36).substring(2, 6).toUpperCase());
+      const randomCodes = new Array(10).fill(1).map(() => randomCode());
+      let hasDbError = false;
+
+      // Loop over codes and attempt to make session
+      for (const join_code of randomCodes) {
+        try {
+          const session = await prisma.session.create({
+            select: defaultSessionSelect,
+            data: {
+              game: { connect: { id: game_id } },
+              join_code,
+              settings,
+            },
+          });
+
+          return session;
+        } catch (error) {
+          if (!(error instanceof Prisma.PrismaClientKnownRequestError && error.code === "P2002")) {
+            console.error(error);
+            hasDbError = true;
+            break;
+          }
+        }
+      }
+
+      // Throw error if failed.
+      throw new TRPCError({
+        code: "INTERNAL_SERVER_ERROR",
+        message: hasDbError ? undefined : "cant generate session code :(",
       });
```

Comment:
We can immediately throw if the db fails for a reason other than the unique code.

See suggestion on how I'd do it (I explicitly left out Prisma from the helper fns as thats an implementation detail that callers shouldn't know abt)

(Disclaimer, I wrote this on my phone in the train, dunno if the syntax is 100% correct)

```suggestion
      // TODO: Extract these to a helper file under /server/utils
      enum DatabaseErrorCode {
  UniqueViolation = "P2002",
      }
      const isDatabaseError = (error: unknown, errorCode: DatabaseErrorCode): boolean => error instanceof Prisma.PrismaClientKnownRequestError && error.code === errorCode;

      // Loop over codes and attempt to make session
      for (const join_code of randomCodes) {
        try {
          const session = await prisma.session.create({
            select: defaultSessionSelect,
            data: {
              game: { connect: { id: game_id } },
              join_code,
              settings,
            },
          });

          return session;
        } catch (error) {
          if (!isDatabaseError(error, DatabaseErrorCode.UniqueViolation)) {
            console.error(error);
            throw new TRPCError({
        code: "INTERNAL_SERVER_ERROR",
        message: "cant generate session code :(",
       });
          }
        }
      }
```